rust/gui-client/src-tauri/src/client/gui.rs

```diff
@@ -820,8 +822,18 @@ async fn run_controller(
                 let new_have_internet = network_changes::check_internet().context("Failed to check for internet")?;
                 if new_have_internet != have_internet {
                     have_internet = new_have_internet;
-                    // TODO: Stop / start / restart connlib as needed here
-                    tracing::info!(?have_internet);
+                    if let Some(session) = controller.session.as_mut() {
+                        tracing::debug!("Internet up/down changed, calling `Session::reconnect`");
+                        session.connlib.reconnect();
+                    }
+                }
+            },
+            r = dns_listener.notified() => {
+                r?;
+                if let Some(session) = controller.session.as_mut() {
+                    tracing::debug!("New DNS resolvers, calling `Session::reconnect`");
+                    // TODO: Pass DNS resolvers to connlib here
+                    session.connlib.reconnect();
```

Comment:
AFAIK `reconnect()` is for network changes and `set_dns(dns_servers: Vec<IpAddr>)` is for DNS. At least that [was my assumption here](https://github.com/firezone/firezone/pull/4133/files#diff-a84e8f62a17ac67f781019e6ac0456567fd18ffa7c13b3248609d78debb6480eR283-R289).

```suggestion
                    tracing::debug!("New DNS resolvers, calling `Session::set_dns`");
                    // TODO: Pass DNS resolvers to connlib here
                    session.connlib.set_dns([]);
```

Granted on some platforms they'll always be called at the same time (macOS, Android?)

Reply:
You're right, I just thought `set_dns` wasn't merged yet. I'll check again

Reply:
Ah, I'm just stubbing it out. I was thinking it was going to be added for this refactor effort

Reply:
Yeah that'll get added in #4163 https://github.com/firezone/firezone/pull/4163/files#diff-0b4003993c22f01feddc49f811f4fde0f8e4111b3c655a6104ee0564efbd5d38R84

Not sure if it makes more sense to keep `reconnect` here or get rid of it and just leave the comments